azure/azevents/azevents.go

```diff
@@ -213,7 +238,91 @@ func (s *Server) RegisterHandlers(r chi.Router) {
 			return
 		}
 
-		// TODO: Add any database persistence and other things we'd want to do after a portfolio was parsed.
-		s.logger.Info("parsed portfolio", zap.String("task_id", string(req.Data.TaskID)))
+		if len(req.Data.Outputs) == 0 {
+			s.logger.Error("webhook response had no processed portfolios", zap.String("event_grid_id", req.ID))
+			http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+			return
+		}
+
+		portfolioIDs := []pacta.PortfolioID{}
+		var ranAt time.Time
+		now := s.now()
+		err := s.db.Transactional(r.Context(), func(tx db.Tx) error {
+			incompleteUploads, err := s.db.IncompleteUploads(tx, req.Data.Request.IncompleteUploadIDs)
+			if err != nil {
+				return fmt.Errorf("reading incomplete uploads: %w", err)
+			}
+			if len(incompleteUploads) == 0 {
+				return fmt.Errorf("no incomplete uploads found for ids: %v", req.Data.Request.IncompleteUploadIDs)
+			}
+			var holdingsDate *pacta.HoldingsDate
+			var ownerID pacta.OwnerID
+			for _, iu := range incompleteUploads {
+				if ownerID == "" {
+					ownerID = iu.Owner.ID
+				} else if ownerID != iu.Owner.ID {
+					return fmt.Errorf("multiple owners found for incomplete uploads: %+v", incompleteUploads)
+				}
+				if iu.HoldingsDate == nil {
+					return fmt.Errorf("incomplete upload %s had no holdings date", iu.ID)
+				}
+				if holdingsDate == nil {
+					holdingsDate = iu.HoldingsDate
+				} else if *holdingsDate != *iu.HoldingsDate {
+					return fmt.Errorf("multiple holdings dates found for incomplete uploads: %+v", incompleteUploads)
+				}
+				ranAt = iu.RanAt
+			}
+			for i, output := range req.Data.Outputs {
+				blobID, err := s.db.CreateBlob(tx, &output.Blob)
+				if err != nil {
+					return fmt.Errorf("creating blob %d: %w", i, err)
+				}
+				portfolioID, err := s.db.CreatePortfolio(tx, &pacta.Portfolio{
+					Owner:        &pacta.Owner{ID: ownerID},
+					Name:         output.Blob.FileName,
+					NumberOfRows: output.LineCount,
+					Blob:         &pacta.Blob{ID: blobID},
+					HoldingsDate: holdingsDate,
+				})
+				if err != nil {
+					return fmt.Errorf("creating portfolio %d: %w", i, err)
+				}
+				portfolioIDs = append(portfolioIDs, portfolioID)
+			}
+			for i, iu := range incompleteUploads {
+				err := s.db.UpdateIncompleteUpload(
+					tx,
+					iu.ID,
+					db.SetIncompleteUploadCompletedAt(now),
+					db.SetIncompleteUploadFailureMessage(""),
+					db.SetIncompleteUploadFailureCode(""))
```

Comment:
Good question - this is in case they were populated before. I realize that's adding code for a state we don't expect to have though, so in hindsight we should (and I will) remove them.